File: server/internal/parser/bpmn_parser.go
package parser

import (
	"encoding/xml"
	"fmt"
	"strings"

	"github.com/bpmn-explorer/server/internal/models"
)

const (
	bpmnNamespace = "http://www.omg.org/spec/BPMN/20100524/MODEL"
	xflowNamespace = "http://example.com/bpmn/xflow-extension"
)

// NodeType 节点类型枚举
const (
	NodeTypeStartEvent              uint32 = 1
	NodeTypeEndEvent                uint32 = 2
	NodeTypeUserTask                uint32 = 3
	NodeTypeServiceTask             uint32 = 4
	NodeTypeExclusiveGateway        uint32 = 5
	NodeTypeParallelGateway         uint32 = 6
	NodeTypeSubProcess              uint32 = 7
	NodeTypeIntermediateEvent       uint32 = 8
	NodeTypeIntermediateCatchEvent  uint32 = 9
	NodeTypeEventBasedGateway       uint32 = 10
	NodeTypeBoundaryEvent           uint32 = 11
	NodeTypeTask                    uint32 = 12 // 普通 Task（抽象任务）
)

// XML 结构体定义，用于解析 BPMN XML
// 注意：encoding/xml 使用本地名称（不带前缀），命名空间通过 XMLName 的 Space 字段处理

type definitions struct {
	XMLName xml.Name `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL definitions"`
	Process process  `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL process"`
}

type process struct {
	XMLName xml.Name `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL process"`
	ID      string   `xml:"id,attr"`
	Name    string   `xml:"name,attr"`
	// 支持多种节点类型
	StartEvents              []startEvent              `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL startEvent"`
	EndEvents                []endEvent                `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL endEvent"`
	Tasks                    []task                    `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL task"`
	UserTasks                []userTask                `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL userTask"`
	ServiceTasks             []serviceTask             `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL serviceTask"`
	ExclusiveGateways        []exclusiveGateway        `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL exclusiveGateway"`
	ParallelGateways         []parallelGateway         `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL parallelGateway"`
	SubProcesses             []subProcess              `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL subProcess"`
	IntermediateCatchEvents  []intermediateCatchEvent  `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL intermediateCatchEvent"`
	EventBasedGateways       []eventBasedGateway       `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL eventBasedGateway"`
	BoundaryEvents           []boundaryEvent           `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL boundaryEvent"`
	SequenceFlows            []sequenceFlow            `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL sequenceFlow"`
	Messages                 []message                 `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL message"`
}

type baseElement struct {
	ID       string   `xml:"id,attr"`
	Name     string   `xml:"name,attr"`
	Incoming []string `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL incoming"`
	Outgoing []string `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL outgoing"`
}

type startEvent struct {
	baseElement
}

type endEvent struct {
	baseElement
}

type task struct {
	baseElement
}

type userTask struct {
	baseElement
}

type serviceTask struct {
	baseElement
	ExtensionElements extensionElements `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL extensionElements"`
}

type extensionElements struct {
	Values []extensionValue `xml:",any"`
}

type extensionValue struct {
	XMLName xml.Name
	Value   string `xml:"value,attr"`
	Content string `xml:",chardata"`
}

type exclusiveGateway struct {
	baseElement
}

type parallelGateway struct {
	baseElement
}

type subProcess struct {
	baseElement
}

type intermediateCatchEvent struct {
	baseElement
}

type eventBasedGateway struct {
	baseElement
}

type boundaryEvent struct {
	baseElement
	AttachedToRef string `xml:"attachedToRef,attr"`
	CancelActivity bool   `xml:"cancelActivity,attr"`
}

type sequenceFlow struct {
	XMLName            xml.Name `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL sequenceFlow"`
	ID                 string    `xml:"id,attr"`
	Name               string    `xml:"name,attr"`
	SourceRef          string    `xml:"sourceRef,attr"`
	TargetRef          string    `xml:"targetRef,attr"`
	ConditionExpression conditionExpression `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL conditionExpression"`
	Priority           uint32    `xml:"priority,attr"`
}

type conditionExpression struct {
	XMLName xml.Name `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL conditionExpression"`
	Type    string   `xml:"type,attr"`
	Content string   `xml:",chardata"`
}

type message struct {
	XMLName xml.Name `xml:"http://www.omg.org/spec/BPMN/20100524/MODEL message"`
	ID      string   `xml:"id,attr"`
	Name    string   `xml:"name,attr"`
}

// ParseBPMN 解析 BPMN XML 内容并返回 WorkflowDefinition
func ParseBPMN(bpmnContent string) (*models.WorkflowDefinition, error) {
	if strings.TrimSpace(bpmnContent) == "" {
		return nil, fmt.Errorf("BPMN content is empty")
	}

	var def definitions
	if err := xml.Unmarshal([]byte(bpmnContent), &def); err != nil {
		return nil, fmt.Errorf("failed to parse XML: %w", err)
	}

	// 检查是否有 process 元素
	if def.Process.ID == "" {
		return nil, fmt.Errorf("missing required element: process")
	}

	wd := &models.WorkflowDefinition{
		Nodes:              make(map[string]models.Node),
		SequenceFlows:      make(map[string]models.SequenceFlow),
		Messages:           make(map[string]models.Message),
		StartEvents:        []string{},
		EndEvents:          []string{},
		VariableDeclarations: []models.VariableDeclaration{},
		AdjacencyList:      make(map[string][]string),
		ReverseAdjacencyList: make(map[string][]string),
	}

	// 解析节点
	parseNodes(&def.Process, wd)

	// 解析序列流
	parseSequenceFlows(&def.Process, wd)

	// 解析消息
	parseMessages(&def.Process, wd)

	// 构建邻接表
	buildAdjacencyLists(wd)

	// 验证 UserTask 约束
	if err := validateUserTaskConstraints(wd); err != nil {
		return nil, err
	}

	// 识别开始和结束事件
	identifyStartAndEndEvents(wd)

	return wd, nil
}

// parseNodes 解析所有节点
func parseNodes(proc *process, wd *models.WorkflowDefinition) {
	// 解析开始事件
	for _, se := range proc.StartEvents {
		node := models.Node{
			Id:                      se.ID,
			Name:                    se.Name,
			Type:                    NodeTypeStartEvent,
			IncomingSequenceFlowIds: se.Incoming,
			OutgoingSequenceFlowIds: se.Outgoing,
			CanFallback:             true,
		}
		wd.Nodes[node.Id] = node
	}

	// 解析结束事件
	for _, ee := range proc.EndEvents {
		node := models.Node{
			Id:                      ee.ID,
			Name:                    ee.Name,
			Type:                    NodeTypeEndEvent,
			IncomingSequenceFlowIds: ee.Incoming,
			OutgoingSequenceFlowIds: ee.Outgoing,
			CanFallback:             true,
		}
		wd.Nodes[node.Id] = node
	}

	// 解析普通任务（Task）
	for _, t := range proc.Tasks {
		node := models.Node{
			Id:                      t.ID,
			Name:                    t.Name,
			Type:                    NodeTypeTask,
			IncomingSequenceFlowIds: t.Incoming,
			OutgoingSequenceFlowIds: t.Outgoing,
			CanFallback:             true,
		}
		wd.Nodes[node.Id] = node
	}

	// 解析用户任务
	for _, ut := range proc.UserTasks {
		node := models.Node{
			Id:                      ut.ID,
			Name:                    ut.Name,
			Type:                    NodeTypeUserTask,
			IncomingSequenceFlowIds: ut.Incoming,
			OutgoingSequenceFlowIds: ut.Outgoing,
			CanFallback:             true,
		}
		wd.Nodes[node.Id] = node
	}

	// 解析服务任务
	for _, st := range proc.ServiceTasks {
		node := models.Node{
			Id:                      st.ID,
			Name:                    st.Name,
			Type:                    NodeTypeServiceTask,
			IncomingSequenceFlowIds: st.Incoming,
			OutgoingSequenceFlowIds: st.Outgoing,
			CanFallback:             true,
		}
		// 从扩展属性中提取业务接口 URL
		if len(st.ExtensionElements.Values) > 0 {
			for _, ext := range st.ExtensionElements.Values {
				// 查找 xflow:Url 或 url 元素
				localName := ext.XMLName.Local
				if localName == "url" || localName == "Url" || strings.HasSuffix(ext.XMLName.Space, "xflow-extension") {
					// 优先使用 value 属性，如果没有则使用内容
					if ext.Value != "" {
						node.BusinessApiUrl = ext.Value
					} else if ext.Content != "" {
						node.BusinessApiUrl = strings.TrimSpace(ext.Content)
					}
					break
				}
			}
		}
		wd.Nodes[node.Id] = node
	}

	// 解析排他网关
	for _, eg := range proc.ExclusiveGateways {
		node := models.Node{
			Id:                      eg.ID,
			Name:                    eg.Name,
			Type:                    NodeTypeExclusiveGateway,
			IncomingSequenceFlowIds: eg.Incoming,
			OutgoingSequenceFlowIds: eg.Outgoing,
			CanFallback:             true,
		}
		wd.Nodes[node.Id] = node
	}

	// 解析并行网关
	for _, pg := range proc.ParallelGateways {
		node := models.Node{
			Id:                      pg.ID,
			Name:                    pg.Name,
			Type:                    NodeTypeParallelGateway,
			IncomingSequenceFlowIds: pg.Incoming,
			OutgoingSequenceFlowIds: pg.Outgoing,
			CanFallback:             true,
		}
		wd.Nodes[node.Id] = node
	}

	// 解析子流程
	for _, sp := range proc.SubProcesses {
		node := models.Node{
			Id:                      sp.ID,
			Name:                    sp.Name,
			Type:                    NodeTypeSubProcess,
			IncomingSequenceFlowIds: sp.Incoming,
			OutgoingSequenceFlowIds: sp.Outgoing,
			CanFallback:             true,
		}
		wd.Nodes[node.Id] = node
	}

	// 解析中间捕获事件
	for _, ice := range proc.IntermediateCatchEvents {
		node := models.Node{
			Id:                      ice.ID,
			Name:                    ice.Name,
			Type:                    NodeTypeIntermediateCatchEvent,
			IncomingSequenceFlowIds: ice.Incoming,
			OutgoingSequenceFlowIds: ice.Outgoing,
			CanFallback:             true,
		}
		wd.Nodes[node.Id] = node
	}

	// 解析事件网关
	for _, ebg := range proc.EventBasedGateways {
		node := models.Node{
			Id:                      ebg.ID,
			Name:                    ebg.Name,
			Type:                    NodeTypeEventBasedGateway,
			IncomingSequenceFlowIds: ebg.Incoming,
			OutgoingSequenceFlowIds: ebg.Outgoing,
			CanFallback:             true,
		}
		wd.Nodes[node.Id] = node
	}

	// 解析边界事件
	for _, be := range proc.BoundaryEvents {
		node := models.Node{
			Id:                      be.ID,
			Name:                    be.Name,
			Type:                    NodeTypeBoundaryEvent,
			IncomingSequenceFlowIds: be.Incoming,
			OutgoingSequenceFlowIds: be.Outgoing,
			AttachedNodeId:          be.AttachedToRef,
			CanFallback:             true,
		}
		wd.Nodes[node.Id] = node
	}
}

// parseSequenceFlows 解析序列流
func parseSequenceFlows(proc *process, wd *models.WorkflowDefinition) {
	for _, sf := range proc.SequenceFlows {
		flow := models.SequenceFlow{
			Id:                sf.ID,
			Name:              sf.Name,
			SourceNodeId:      sf.SourceRef,
			TargetNodeId:      sf.TargetRef,
			ConditionExpression: strings.TrimSpace(sf.ConditionExpression.Content),
			Priority:          sf.Priority,
		}
		wd.SequenceFlows[flow.Id] = flow
	}
}

// parseMessages 解析消息
func parseMessages(proc *process, wd *models.WorkflowDefinition) {
	for _, msg := range proc.Messages {
		message := models.Message{
			Id:   msg.ID,
			Name: msg.Name,
		}
		wd.Messages[message.Id] = message
	}
}

// buildAdjacencyLists 构建邻接表
func buildAdjacencyLists(wd *models.WorkflowDefinition) {
	// 初始化邻接表
	for nodeID := range wd.Nodes {
		wd.AdjacencyList[nodeID] = []string{}
		wd.ReverseAdjacencyList[nodeID] = []string{}
	}

	// 根据序列流构建邻接表
	for _, flow := range wd.SequenceFlows {
		sourceID := flow.SourceNodeId
		targetID := flow.TargetNodeId

		// 检查源节点和目标节点是否存在
		if _, exists := wd.Nodes[sourceID]; !exists {
			continue
		}
		if _, exists := wd.Nodes[targetID]; !exists {
			continue
		}

		// 添加到正向邻接表
		wd.AdjacencyList[sourceID] = append(wd.AdjacencyList[sourceID], targetID)

		// 添加到反向邻接表
		wd.ReverseAdjacencyList[targetID] = append(wd.ReverseAdjacencyList[targetID], sourceID)
	}
}

// validateUserTaskConstraints 验证 UserTask 的 outgoing 连线约束
// 确保所有 UserTask 的 outgoing 连线都从 BoundaryEvent 出发
func validateUserTaskConstraints(wd *models.WorkflowDefinition) error {
	// 1. 构建 BoundaryEvent 索引：attachedNodeId -> []boundaryEventIDs
	boundaryEvents := make(map[string][]string)
	for nodeID, node := range wd.Nodes {
		if node.Type == NodeTypeBoundaryEvent {
			if node.AttachedNodeId == "" {
				return fmt.Errorf("BoundaryEvent %s missing attachedToRef", nodeID)
			}
			boundaryEvents[node.AttachedNodeId] = append(
				boundaryEvents[node.AttachedNodeId], nodeID)
		}
	}

	// 2. 验证每个 UserTask
	for nodeID, node := range wd.Nodes {
		if node.Type != NodeTypeUserTask {
			continue
		}

		if len(node.OutgoingSequenceFlowIds) == 0 {
			continue // 没有 outgoing，跳过
		}

		// 检查每条 outgoing 连线的 sourceRef
		for _, flowID := range node.OutgoingSequenceFlowIds {
			flow, exists := wd.SequenceFlows[flowID]
			if !exists {
				return fmt.Errorf("SequenceFlow %s not found", flowID)
			}

			// 验证失败：连线直接从 UserTask 出发
			if flow.SourceNodeId == nodeID {
				return fmt.Errorf(
					"UserTask %s has direct outgoing flow %s. "+
						"All outgoing flows from UserTask must originate from BoundaryEvent",
					nodeID, flowID)
			}
		}

		// 检查是否有 BoundaryEvent
		if len(boundaryEvents[nodeID]) == 0 {
			return fmt.Errorf(
				"UserTask %s has outgoing flows but no BoundaryEvent attached",
				nodeID)
		}
	}

	return nil
}

// identifyStartAndEndEvents 识别开始和结束事件
func identifyStartAndEndEvents(wd *models.WorkflowDefinition) {
	for nodeID, node := range wd.Nodes {
		switch node.Type {
		case NodeTypeStartEvent:
			wd.StartEvents = append(wd.StartEvents, nodeID)
		case NodeTypeEndEvent:
			wd.EndEvents = append(wd.EndEvents, nodeID)
		}
	}
}

File: server-nodejs/src/parser/bpmnParser.ts
import { XMLParser } from 'fast-xml-parser';
import { WorkflowDefinition, Node, SequenceFlow, NodeType } from '../services/workflowEngineService';

interface ExtensionValue {
  '@_value'?: string;
  '#text'?: string;
}

interface ExtensionElements {
  'xflow:url'?: ExtensionValue;
  'xflow:Url'?: ExtensionValue;
  url?: ExtensionValue;
  Url?: ExtensionValue;
}

interface BaseElement {
  '@_id': string;
  '@_name'?: string;
  'bpmn:incoming'?: string | string[];
  'bpmn:outgoing'?: string | string[];
}

interface ServiceTask extends BaseElement {
  'bpmn:extensionElements'?: ExtensionElements;
}

interface BoundaryEvent extends BaseElement {
  '@_attachedToRef': string;
  '@_cancelActivity'?: string;
}

interface SequenceFlowXML {
  '@_id': string;
  '@_name'?: string;
  '@_sourceRef': string;
  '@_targetRef': string;
  '@_priority'?: number;
  'bpmn:conditionExpression'?: {
    '@_type'?: string;
    '#text'?: string;
  };
}

interface Process {
  '@_id': string;
  '@_name'?: string;
  'bpmn:startEvent'?: BaseElement | BaseElement[];
  'bpmn:endEvent'?: BaseElement | BaseElement[];
  'bpmn:task'?: BaseElement | BaseElement[];
  'bpmn:userTask'?: BaseElement | BaseElement[];
  'bpmn:serviceTask'?: ServiceTask | ServiceTask[];
  'bpmn:exclusiveGateway'?: BaseElement | BaseElement[];
  'bpmn:parallelGateway'?: BaseElement | BaseElement[];
  'bpmn:subProcess'?: BaseElement | BaseElement[];
  'bpmn:intermediateCatchEvent'?: BaseElement | BaseElement[];
  'bpmn:eventBasedGateway'?: BaseElement | BaseElement[];
  'bpmn:boundaryEvent'?: BoundaryEvent | BoundaryEvent[];
  'bpmn:sequenceFlow'?: SequenceFlowXML | SequenceFlowXML[];
  'bpmn:message'?: { '@_id': string; '@_name'?: string } | { '@_id': string; '@_name'?: string }[];
}

interface Definitions {
  'bpmn:definitions': {
    'bpmn:process': Process | Process[];
  };
}

/**
 * Parse BPMN XML content and return WorkflowDefinition
 */
export function parseBPMN(bpmnContent: string): WorkflowDefinition {
  if (!bpmnContent || bpmnContent.trim() === '') {
    throw new Error('BPMN content is empty');
  }

  const parser = new XMLParser({
    ignoreAttributes: false,
    attributeNamePrefix: '@_',
    parseAttributeValue: true,
    trimValues: true,
  });

  const result = parser.parse(bpmnContent) as Definitions;

  // Check if process element exists
  const processData = result['bpmn:definitions']?.['bpmn:process'];
  if (!processData) {
    throw new Error('Missing required element: process');
  }

  // Handle single or multiple processes (take first one)
  const process = Array.isArray(processData) ? processData[0] : processData;

  if (!process['@_id']) {
    throw new Error('Process element missing id attribute');
  }

  const wd: WorkflowDefinition = {
    processId: process['@_id'],
    nodes: {},
    sequenceFlows: {},
    startEvents: [],
    adjacencyList: {},
    reverseAdjacencyList: {},
  };

  // Parse nodes
  parseNodes(process, wd);

  // Parse sequence flows
  parseSequenceFlows(process, wd);

  // Build adjacency lists
  buildAdjacencyLists(wd);

  // Validate UserTask constraints
  validateUserTaskConstraints(wd);

  // Identify start and end events
  identifyStartAndEndEvents(wd);

  return wd;
}

/**
 * Convert element data to array if single element
 */
function toArray<T>(data: T | T[] | undefined): T[] {
  if (!data) return [];
  return Array.isArray(data) ? data : [data];
}

/**
 * Convert incoming/outgoing to string array
 */
function toStringArray(data: string | string[] | undefined): string[] {
  if (!data) return [];
  return Array.isArray(data) ? data : [data];
}

/**
 * Parse all nodes from process
 */
function parseNodes(process: Process, wd: WorkflowDefinition): void {
  // Parse start events
  toArray(process['bpmn:startEvent']).forEach((se) => {
    const node: Node = {
      id: se['@_id'],
      name: se['@_name'] || '',
      type: NodeType.START_EVENT,
      incomingSequenceFlowIds: toStringArray(se['bpmn:incoming']),
      outgoingSequenceFlowIds: toStringArray(se['bpmn:outgoing']),
      canFallback: true,
    };
    wd.nodes[node.id] = node;
  });

  // Parse end events
  toArray(process['bpmn:endEvent']).forEach((ee) => {
    const node: Node = {
      id: ee['@_id'],
      name: ee['@_name'] || '',
      type: NodeType.END_EVENT,
      incomingSequenceFlowIds: toStringArray(ee['bpmn:incoming']),
      outgoingSequenceFlowIds: toStringArray(ee['bpmn:outgoing']),
      canFallback: true,
    };
    wd.nodes[node.id] = node;
  });

  // Parse tasks
  toArray(process['bpmn:task']).forEach((t) => {
    const node: Node = {
      id: t['@_id'],
      name: t['@_name'] || '',
      type: NodeType.TASK,
      incomingSequenceFlowIds: toStringArray(t['bpmn:incoming']),
      outgoingSequenceFlowIds: toStringArray(t['bpmn:outgoing']),
      canFallback: true,
    };
    wd.nodes[node.id] = node;
  });

  // Parse user tasks
  toArray(process['bpmn:userTask']).forEach((ut) => {
    const node: Node = {
      id: ut['@_id'],
      name: ut['@_name'] || '',
      type: NodeType.USER_TASK,
      incomingSequenceFlowIds: toStringArray(ut['bpmn:incoming']),
      outgoingSequenceFlowIds: toStringArray(ut['bpmn:outgoing']),
      canFallback: true,
    };
    wd.nodes[node.id] = node;
  });

  // Parse service tasks
  toArray(process['bpmn:serviceTask']).forEach((st) => {
    const node: Node = {
      id: st['@_id'],
      name: st['@_name'] || '',
      type: NodeType.SERVICE_TASK,
      incomingSequenceFlowIds: toStringArray(st['bpmn:incoming']),
      outgoingSequenceFlowIds: toStringArray(st['bpmn:outgoing']),
      canFallback: true,
    };

    // Extract business API URL from extension elements
    if (st['bpmn:extensionElements']) {
      const ext = st['bpmn:extensionElements'];
      const urlExt = ext['xflow:url'] || ext['xflow:Url'] || ext['url'] || ext['Url'];
      if (urlExt) {
        node.businessApiUrl = urlExt['@_value'] || urlExt['#text'] || '';
      }
    }

    wd.nodes[node.id] = node;
  });

  // Parse exclusive gateways
  toArray(process['bpmn:exclusiveGateway']).forEach((eg) => {
    const node: Node = {
      id: eg['@_id'],
      name: eg['@_name'] || '',
      type: NodeType.EXCLUSIVE_GATEWAY,
      incomingSequenceFlowIds: toStringArray(eg['bpmn:incoming']),
      outgoingSequenceFlowIds: toStringArray(eg['bpmn:outgoing']),
      canFallback: true,
    };
    wd.nodes[node.id] = node;
  });

  // Parse parallel gateways
  toArray(process['bpmn:parallelGateway']).forEach((pg) => {
    const node: Node = {
      id: pg['@_id'],
      name: pg['@_name'] || '',
      type: NodeType.PARALLEL_GATEWAY,
      incomingSequenceFlowIds: toStringArray(pg['bpmn:incoming']),
      outgoingSequenceFlowIds: toStringArray(pg['bpmn:outgoing']),
      canFallback: true,
    };
    wd.nodes[node.id] = node;
  });

  // Parse intermediate catch events
  toArray(process['bpmn:intermediateCatchEvent']).forEach((ice) => {
    const node: Node = {
      id: ice['@_id'],
      name: ice['@_name'] || '',
      type: NodeType.INTERMEDIATE_CATCH_EVENT,
      incomingSequenceFlowIds: toStringArray(ice['bpmn:incoming']),
      outgoingSequenceFlowIds: toStringArray(ice['bpmn:outgoing']),
      canFallback: true,
    };
    wd.nodes[node.id] = node;
  });

  // Parse event-based gateways
  toArray(process['bpmn:eventBasedGateway']).forEach((ebg) => {
    const node: Node = {
      id: ebg['@_id'],
      name: ebg['@_name'] || '',
      type: NodeType.EVENT_BASED_GATEWAY,
      incomingSequenceFlowIds: toStringArray(ebg['bpmn:incoming']),
      outgoingSequenceFlowIds: toStringArray(ebg['bpmn:outgoing']),
      canFallback: true,
    };
    wd.nodes[node.id] = node;
  });

  // Parse boundary events
  toArray(process['bpmn:boundaryEvent']).forEach((be) => {
    const node: Node = {
      id: be['@_id'],
      name: be['@_name'] || '',
      type: NodeType.BOUNDARY_EVENT,
      incomingSequenceFlowIds: toStringArray(be['bpmn:incoming']),
      outgoingSequenceFlowIds: toStringArray(be['bpmn:outgoing']),
      attachedNodeId: be['@_attachedToRef'],
      canFallback: true,
    };
    wd.nodes[node.id] = node;
  });
}

/**
 * Parse sequence flows
 */
function parseSequenceFlows(process: Process, wd: WorkflowDefinition): void {
  toArray(process['bpmn:sequenceFlow']).forEach((sf) => {
    const flow: SequenceFlow = {
      id: sf['@_id'],
      sourceNodeId: sf['@_sourceRef'],
      targetNodeId: sf['@_targetRef'],
      conditionExpression: sf['bpmn:conditionExpression']?.['#text']?.trim() || '',
    };
    wd.sequenceFlows[flow.id] = flow;
  });
}

/**
 * Build adjacency lists for graph traversal
 */
function buildAdjacencyLists(wd: WorkflowDefinition): void {
  // Initialize adjacency lists
  Object.keys(wd.nodes).forEach((nodeId) => {
    wd.adjacencyList[nodeId] = [];
    wd.reverseAdjacencyList[nodeId] = [];
  });

  // Build adjacency lists from sequence flows
  Object.values(wd.sequenceFlows).forEach((flow) => {
    const sourceId = flow.sourceNodeId;
    const targetId = flow.targetNodeId;

    // Check if source and target nodes exist
    if (!wd.nodes[sourceId] || !wd.nodes[targetId]) {
      return;
    }

    // Add to forward adjacency list
    wd.adjacencyList[sourceId].push(targetId);

    // Add to reverse adjacency list
    wd.reverseAdjacencyList[targetId].push(sourceId);
  });
}

/**
 * Validate UserTask constraints
 * Ensures all UserTask outgoing flows originate from BoundaryEvent
 */
function validateUserTaskConstraints(wd: WorkflowDefinition): void {
  // Build BoundaryEvent index: attachedNodeId -> boundaryEventIds
  const boundaryEvents: Record<string, string[]> = {};
  Object.entries(wd.nodes).forEach(([nodeId, node]) => {
    if (node.type === NodeType.BOUNDARY_EVENT) {
      if (!node.attachedNodeId) {
        throw new Error(`BoundaryEvent ${nodeId} missing attachedToRef`);
      }
      if (!boundaryEvents[node.attachedNodeId]) {
        boundaryEvents[node.attachedNodeId] = [];
      }
      boundaryEvents[node.attachedNodeId].push(nodeId);
    }
  });

  // Validate each UserTask
  Object.entries(wd.nodes).forEach(([nodeId, node]) => {
    if (node.type !== NodeType.USER_TASK) {
      return;
    }

    if (node.outgoingSequenceFlowIds.length === 0) {
      return; // No outgoing flows, skip
    }

    // Check each outgoing flow's sourceRef
    node.outgoingSequenceFlowIds.forEach((flowId) => {
      const flow = wd.sequenceFlows[flowId];
      if (!flow) {
        throw new Error(`SequenceFlow ${flowId} not found`);
      }

      // Validation fails: flow originates directly from UserTask
      if (flow.sourceNodeId === nodeId) {
        throw new Error(
          `UserTask ${nodeId} has direct outgoing flow ${flowId}. ` +
            `All outgoing flows from UserTask must originate from BoundaryEvent`
        );
      }
    });

    // Check if BoundaryEvent exists
    if (!boundaryEvents[nodeId] || boundaryEvents[nodeId].length === 0) {
      throw new Error(
        `UserTask ${nodeId} has outgoing flows but no BoundaryEvent attached`
      );
    }
  });
}

/**
 * Identify start and end events
 */
function identifyStartAndEndEvents(wd: WorkflowDefinition): void {
  Object.entries(wd.nodes).forEach(([nodeId, node]) => {
    if (node.type === NodeType.START_EVENT) {
      wd.startEvents.push(nodeId);
    }
  });
}
